### PR TITLE
Online DDL: improve retry of vreplication errors with `vitess` `ALTER TABLE` migrations

### PR DESCRIPTION
**Pull request by shlomi-noach**
## Description

PEr https://github.com/vitessio/vitess/issues/12322, Online DDL gives up too soon in some scenarios of `vitess` migrations; the scenarios are where there's a temporary, recoverable error on `vreplication`'s side; Online DDL chooses to terminate the migration rather than wait for `vreplication` to recover.

The problem is that it's hard for Online DDL to know when a VReplication error is terminal or recoverable. 

The main change in this PR is for the Online DDL scheduler to allow retry/timeout on `VReplication` errors, by using `LastError`, refactored in https://github.com/vitessio/vitess/pull/12321. This in effect allows up to `10min` of retries (with at least one check per minute) before Online DDL gives up on the vreplication stream.

So instead of analyzing specific errors, we just wait/retry on `vreplication` up to a timeout.

Other changes in this PR, while we're here:

- Proper cleanup of `ownedRunningMigrations` upon `Open()`. This may have also contributed on `onlineddl_vrepl` errors, though I'm not sure, and the cleanup presented here is required either way.
- Change `isOpen` from `bool` to `int64` and use with `atomic.LoadInt64()/StoreInt64()` as the reads/writes to this variables may be concurrent.
- More information in the logs when adopting a migration from a previous failed/demoted tablet
- More information in the logs when running `gcArtifacts`

## Related Issue(s)

Fixes https://github.com/vitessio/vitess/issues/12322
Extends https://github.com/vitessio/vitess/pull/12321 (not merged yet)
Depends on https://github.com/vitessio/vitess/pull/12327

## Checklist

-   [ ] "Backport to:" labels have been added if this change should be back-ported
-   [ ] Tests were added or are not required
-   [ ] Documentation was added or is not required

### CODE REVIEW

**Comment by shlomi-noach** [comment]
So on the flip side, the `onlineddl_vrepl_suite` tests fail. The failures are for the `fail-*` tests; that is, these are tests that expect to fail. But the new wait/retry logic just... keeps waiting. The failure truly are terminal, but Online DDL waits for some `10min` to be sure. However, the tests cannot each wait for `10min`.

Gonna have to do some thinking here.

**Comment by shlomi-noach** [comment]
All right, now trying to differentiate and identify those workflows which `vreplication` already admits to be completely failed.

**Comment by shlomi-noach** [comment]
Not good enough. For example, the `fail-cannot-be-null` test causes this scenario in VReplication:

```

*************************** 1. row ***************************
                   id: 1
             workflow: 6a363eb7_ab59_11ed_b253_0a43f95f28a3
               source: keyspace:"commerce" shard:"0" filter:{rules:{match:"_6a363eb7_ab59_11ed_b253_0a43f95f28a3_20230213044645_vrepl" filter:"select `id` as `id`, `val` as `val`, `ts` as `ts` from `onlineddl_test`" source_unique_key_columns:"id" target_unique_key_columns:"id" sour
ce_unique_key_target_columns:"id"}}
                  pos: MySQL56/27770cc8-ab59-11ed-b694-0a43f95f28a3:1-68
             stop_pos: NULL
              max_tps: 9223372036854775807
  max_replication_lag: 9223372036854775807
                 cell:
         tablet_types: in_order:REPLICA,PRIMARY
         time_updated: 1676263623
transaction_timestamp: 1676263623
                state: Copying
              message: task error: failed inserting rows: Column 'val' cannot be null (errno 1048) (sqlstate 23000) during query: insert into _6a363eb7_ab59_11ed_b253_0a43f95f28a3_20230213044645_vrepl(id,val,ts) values (1,2,'2023-02-13 04:46:16'), (2,3,'2023-02-13 04:46:16'), (3,nu
ll,'2023-02-13 04:46:16')
              db_name: vt_commerce
          rows_copied: 0
                 tags:
       time_heartbeat: 1676263623
        workflow_type: 5
       time_throttled: 0
  component_throttled:
    workflow_sub_type: 0
 defer_secondary_keys: 0
```
And `vreplication` still does not see this as terminal. Looking further.

**Comment by shlomi-noach** [comment]
Error code `1048` is already listed as "unrecoverable error" in `vreplication`, here: https://github.com/vitessio/vitess/blob/c4292b7c560766275e22be8105cf95d9d5afe7b8/go/vt/vttablet/tabletmanager/vreplication/utils.go#L136

So I'm not sure why VReplication itself did not completely bail out?

**Comment by shlomi-noach** [comment]
see https://github.com/vitessio/vitess/issues/12326 ; this PR will continue to fail CI until https://github.com/vitessio/vitess/issues/12326 is resolved.

**Comment by shlomi-noach** [comment]
Depends on https://github.com/vitessio/vitess/pull/12327

**Review by mattlord — APPROVED** [review]
LGTM! I had some minor comments that you can resolve as you think best.

**Review by mattlord — COMMENTED** [review]
This did seem to cause some valid test failures/breakage in the `onlineddl_vrepl_suite` workflows so we'll need to fix that.

File: go/vt/vttablet/onlineddl/executor.go [hunk 1]
```diff
@@ -208,11 +208,12 @@ type Executor struct {
 	// - be terminated (example: pt-osc migration gone rogue, process still running even as the migration failed)
 	// The Executor auto-reviews the map and cleans up migrations thought to be running which are not running.
 	ownedRunningMigrations        sync.Map
+	vreplicationLastError         map[string]*vterrors.LastError
 	tickReentranceFlag            int64
 	reviewedRunningMigrationsFlag bool
 
 	ticks             *timer.Timer
-	isOpen            bool
+	isOpen            int64
```
[the remark this answers is not in the document]
Comment (shlomi-noach) [reply go/vt/vttablet/onlineddl/executor.go:216]: Oh interesting, I did not realize that!

File: go/vt/vttablet/onlineddl/executor.go [hunk 2]
```diff
@@ -3853,10 +3883,11 @@ func (e *Executor) gcArtifacts(ctx context.Context) error {
...
 			log.Infof("Executor.gcArtifacts: will GC artifact %s for migration %s", artifactTable, uuid)
 			t := timeNow.Add(time.Duration(i) * time.Second).UTC()
-			if err := e.gcArtifactTable(ctx, artifactTable, uuid, t); err != nil {
-				return err
+			toTableName, err := e.gcArtifactTable(ctx, artifactTable, uuid, t)
+			if err != nil {
+				return vterrors.Wrapf(err, "in gcArtifacts() for %s", artifactTable)
 			}
-			log.Infof("Executor.gcArtifacts: renamed away artifact %s", artifactTable)
+			log.Infof("Executor.gcArtifacts: renamed away artifact %s into %s", artifactTable, toTableName)
```
[the remark this answers is not in the document]
Comment (shlomi-noach) [reply go/vt/vttablet/onlineddl/executor.go:3890]: Thank you. Please consider using the GitHub "suggest" feature so I can commit directly from your comment.

File: go/vt/vttablet/onlineddl/vrepl.go [hunk 3]
```diff
@@ -80,15 +80,15 @@ func (v *VReplStream) isRunning() bool {
...
 
-// isFailed() returns true when the workflow has failed and will not retry
-func (v *VReplStream) isFailed() bool {
+// hasError() returns true when the workflow has failed and will not retry
+func (v *VReplStream) hasError() (hasErr bool, isTerminal bool, message string) {
 	switch {
 	case v.state == binlogplayer.BlpError:
-		return true
+		return true, true, v.message
 	case strings.Contains(strings.ToLower(v.message), "error"):
```
[the remark this answers is not in the document]
Comment (shlomi-noach) [reply go/vt/vttablet/onlineddl/vrepl.go:88]: I feel good with this logic. Much better than before.

File: go/vt/vttablet/onlineddl/vrepl.go [hunk 4]
```diff
@@ -80,15 +80,15 @@ func (v *VReplStream) isRunning() bool {
...
 }
 
-// isFailed() returns true when the workflow has failed and will not retry
-func (v *VReplStream) isFailed() bool {
+// hasError() returns true when the workflow has failed and will not retry
+func (v *VReplStream) hasError() (hasErr bool, isTerminal bool, message string) {
 	switch {
 	case v.state == binlogplayer.BlpError:
-		return true
+		return true, true, v.message
```
[the remark this answers is not in the document]
Comment (shlomi-noach) [reply go/vt/vttablet/onlineddl/vrepl.go]: Done! And this simplifies the code. Please take another look.

File: go/vt/vttablet/onlineddl/vrepl.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (shlomi-noach) [reply go/vt/vttablet/onlineddl/vrepl.go:89]: Simplified, per https://github.com/vitessio/vitess/pull/12323#discussion_r1105350189

**Comment by shlomi-noach** [comment]
> This did seem to cause some valid test failures/breakage in the onlineddl_vrepl_suite workflows so we'll need to fix that.

Yes, as mentioned in https://github.com/vitessio/vitess/pull/12323#issuecomment-1427412397 and https://github.com/vitessio/vitess/pull/12323#issuecomment-1427495932 ; the errors will resolve when we merge #12327

**Review by rohit-nayak-ps — APPROVED** [review]
lgtm

**Comment by shlomi-noach** [comment]
merged https://github.com/vitessio/vitess/pull/12327 ; now this PR should pass all tests

**Comment by shlomi-noach** [comment]
`Upgrade Downgrade` tests are flaky and taking very long; I'm going to merge this PR now.

**Comment by shlomi-noach** [comment]
I will cherry pick this PR manually, along with https://github.com/vitessio/vitess/pull/12327

**Pull request merged**